From 251c9ce7ba02b2b2eb938fd4de6c06ee927ef5f0 Mon Sep 17 00:00:00 2001
From: Daniel Leung <daniel.leung@intel.com>
Date: Sun, 31 Aug 2014 17:33:45 -0700
Subject: [PATCH 2/2] input: dollar_cove_powerbtn: enable device PM wakeup

When the power button is pressed during execution of suspend path,
the system needs to abort suspend. This is accomplished by marking
this device as a wakeup device, and emits a wakeup event when
interrupt comes in.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>
---
 drivers/input/keyboard/dollar_cove_powerbtn.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/input/keyboard/dollar_cove_powerbtn.c b/drivers/input/keyboard/dollar_cove_powerbtn.c
index 37c7028..351003a 100644
--- a/drivers/input/keyboard/dollar_cove_powerbtn.c
+++ b/drivers/input/keyboard/dollar_cove_powerbtn.c
@@ -25,6 +25,7 @@
 #include <linux/platform_device.h>
 #include <linux/input.h>
 #include <linux/io.h>
+#include <linux/pm_runtime.h>
 
 #define DRIVER_NAME "dollar_cove_power_button"
 
@@ -33,10 +34,15 @@ static int irq_press, irq_release;
 
 static irqreturn_t pb_isr(int irq, void *dev_id)
 {
+	struct platform_device *pdev = dev_id;
+
 	input_event(pb_input, EV_KEY, KEY_POWER, irq == irq_press);
 	input_sync(pb_input);
 	pr_info("[%s] power button %s\n", pb_input->name,
 			irq == irq_press ? "pressed" : "released");
+
+	pm_wakeup_event(&pdev->dev, 0);
+
 	return IRQ_HANDLED;
 }
 
@@ -88,6 +94,11 @@ static int pb_probe(struct platform_device *pdev)
 		return ret;
 	}
 
+	ret = device_init_wakeup(&pdev->dev, 1);
+	if (ret)
+		dev_warn(&pdev->dev,
+			"cannot make this a wakeup device: %d\n", ret);
+
 	return 0;
 }
 
-- 
1.8.3.2

